### ISSUE TITLE
file locking on mac errors.  I'm not sure why.  

### ISSUE THREAD

**Issue by pdebruic** [ApptiveGrid/MoniBase#124]
This is with v0.6 of MoniBase.  On Macos 12.3.1

Running 

```
db:=OmniBase createOn: '/tmp/omnibase/'.
```

fails with `ODBLockingConflict`

Could be a Mac specific permissions issue but the `omnibase.odl` file is created.   

Kind of out of my depth but is there a way to see the `errno` generated by the FFI call to `fcntl` ?  

What I've found is the man page for fcntl gives the flock struct as 

```
struct flock {
                 off_t       l_start;    /* starting offset */
                 off_t       l_len;      /* len = 0 means until end of file */
                 pid_t       l_pid;      /* lock owner */
                 short       l_type;     /* lock type: read/write, etc. */
                 short       l_whence;   /* type of l_start */
             };
```

types.h gives the type for `off_t` and `pid_t` as

```
 typedef int64_t off_t;
 typedef int     pid_t;
```

Which I think means signed int 64 and signed int 32 (because they also describe a u_int64_t and u_int)

In `BSDFlock>>#fieldsDesc`  the struct is described as 

```
fieldsDesc
	"self rebuildFieldAccessors"

	^ #(
	__off64_t l_start;
	__off64_t l_len;
	__pid_t l_pid;	
	int l_type;
	int l_whence;
	)
```

And `__off64_t` is defined as `FFIUInt64` and `__pid_t` is `FFIUInt32`

So basically I don't know if `fieldsDesc` is in fact correct  and its something to do with my machine or if `fieldsDesc` is wrong. 

 Is there any information I could give to help diagnose why its not working?

**Comment by pdebruic** [comment]
oh and `int64_t` in types.h is defined as `quad_t`  if that matters.......

**Comment by MarcusDenker** [comment]
I tried it (Pharo9, Some MacOS, 0.6). And it works... but my user is admin I think.

Did you try in another directory outside of tmp?

**Comment by MarcusDenker** [comment]
One idea (I did not test it): I start Pharo from the command line. I think if you open it using the finder (by clicking on it), it could be that it runs in a kind of sandbox. This is a MacOS feature of the last versions, I think. And there it might not have access to /tmp/

**Comment by MarcusDenker** [comment]
I sadly do not have an ARM mac yet. Thus it is X86

**Comment by pdebruic** [comment]
OK.  Is there a way to get the `errno` that is set by `fcntl` when acquiring the lock fails?

**Comment by pdebruic** [comment]
Oh!  Can paste what the `flock` struct is described as is your fcntl on x86 by running a 

`man fcntl` 

and scrolling down to where the `F_GETLK` part is ?

Maybe they changed it between architectures.....

**Comment by pdebruic** [comment]
Thanks.  Looks like that didn't change between architectures. 

Just as an FYI  - I installed a clean install of macos 12.3.1 for the arm M1 on a vm using https://mac.getutm.app.  

Then tried to create an OmniBase but it gets the same `ODBLockingConflict`.   

So I think its not my machine that has the problem but something more general.

**Comment by MarcusDenker** [comment]
- How do start Pharo? I used the command line.
- Did you try something else than /tmp ?

**Comment by pdebruic** [comment]
Hi - 

Yes the problem happens with the M1 vm for me when starting Pharo both from the command line (via zeroconf) and the launcher.  This is with Pharo 9.  I will check Pharo 10 to see if its different.  

I have tested attempting to create the the omnibase database next to the image, in my home folder, & in '/tmp'.   

I tested that in both my machine and a macos M1 vm I set up just to test omnibase db creation.

If I use the x86 vm & image on the M1 machine the database is created and seems to work fine.  

If I use the linux ARM vm & image  in an ubuntu 22.04 vm on my M1 mac the database is created and seems to work fine.  

I don't know how to interpret the file handle that the FilePlugin creates to see if its different between the x86 vm and the M1 arm VM but will get them and post them here if they're different in any way.  

Thanks for following up.

**Comment by MarcusDenker** [comment]
This means this is very likely due to some difference X86 vs ARM and on the mac (I edited the title a little)

**Comment by MarcusDenker** [comment]
I have now a M1 and will try to reproduce the problem

**Comment by MarcusDenker** [comment]
So: yes, file locking does not work with the M1 VM, if I use the X86 VM (on the M1 under Rosetta) it works.
(not that helpful, but simplifies me switching to the M1 laptop).

**Comment by MarcusDenker** [comment]
The errno can be read like this:

```
(ExternalAddress loadSymbol: 'errno' ) unsignedLongAt: 1
````

If I add a transcript output in  BSDFLock class>>#lock:from:to:exclusive:, I get 14. That is EFAULT.

The macos man page does not mention that, but linus says:

```
 EFAULT   lock is outside your accessible address space.
```

**Comment by MarcusDenker** [comment]
The lock is from the FILE datastructure, BinaryFileStream>>#fileHandle is:

```
fileHandle

	^ handle pointerAt: 9

```
Maybe the pointer is wrong? but then, I would expect BSDFLock class>>#fileno: to fail first...

**Comment by MarcusDenker** [comment]
I compiled a VM for M1 from the latest sources (there are some FFI related fixes that soon will be in the release).

Sadly, these do not fix our problem.

So my theory is that the "handle pointerAt: 9" is not returning a valid file handle...

**Comment by MarcusDenker** [comment]
That took a while... but I found it: on the M1 it is important to use the "fixedArgumentCount:" variation when calling variadic arguments, so that the right calling convention can be used

```
fcntl: fd command: cmd struct: struct
	^ self 
		ffiCall: #(int fcntl(int fd, int cmd, FLock *struct)) 
		library: LibC
		fixedArgumentCount:2
```
